tests/pytests/unit/fileserver/test_s3fs.py

```diff
@@ -1,19 +1,151 @@
+import os
+
+import boto3
 import pytest
+import yaml
+
+# moto must be imported before boto3
+from moto import mock_s3
 
 import salt.fileserver.s3fs as s3fs
+import salt.utils.s3
+from tests.support.mock import patch
 
 
 @pytest.fixture
-def configure_loader_modules(tmp_path):
+def bucket():
+    with mock_s3():
+        yield "mybucket"
+
+
+@pytest.fixture(scope="module")
+def aws_creds():
+    return {
+        "aws_access_key_id": "testing",
+        "aws_secret_access_key": "testing",
+        "aws_session_token": "testing",
+        "region_name": "us-east-1",
+    }
+
+
+@pytest.fixture(scope="function")
+def configure_loader_modules(tmp_path, bucket):
     opts = {
         "cachedir": tmp_path,
+        "s3.buckets": {"base": [bucket]},
+        "s3.location": "us-east-1",
     }
-    return {s3fs: {"__opts__": opts}}
+    utils = {"s3.query": salt.utils.s3.query}
+
+    return {s3fs: {"__opts__": opts, "__utils__": utils}}
+
+
+@pytest.fixture(scope="function")
+def s3(bucket, aws_creds):
+    def _s3():
+        conn = boto3.client("s3", **aws_creds)
+        conn.create_bucket(Bucket=bucket)
+        return conn
+
+    return _s3
```

Comment:
question: Why does this fixture return a factory instead of the object? Doing the setup in the fixture function itself could DRY up the tests and is how `pytest` envisions fixtures afaik.

```suggestion
@pytest.fixture
def files_present():
    return {
        "top.sls": {"content": yaml.dump({"base": {"*": ["foo"]}})},
        "foo.sls": {"content": yaml.dump({"nginx": {"pkg.installed": []}})},
        "files/nginx.conf": {"content": "server {}"},
    }


@pytest.fixture(scope="function")
def s3(bucket, aws_creds, files_present):
    conn = boto3.client("s3", **aws_creds)
    conn.create_bucket(Bucket=bucket)
    make_keys(bucket, conn, files_present)
    s3fs.update()
    verify_cache(bucket, files_present)
    return conn
```

Reply:
Actually you're right! It was necessary before I got the bucket being created in a fixture as well, but it's not any more. I changed it to just a function scoped fixture instead of a factory.

However, I think making the particular test files that get uploaded into a fixture takes it a bit too far, since they are particular test inputs. They happen to be the same here, but even in #65614, which will follow this one, the files change slightly to test another case.

Reply:
You could use indirect parametrization in that case, but admittedly that gets increasingly complex...

```python
@pytest.fixture(params=("default_files",))
def files_present(request):
    return request.getfixturevalue(request.param)


@pytest.fixture
def default_files():
    return {
        "top.sls": {"content": yaml.dump({"base": {"*": ["foo"]}})},
        "foo.sls": {"content": yaml.dump({"nginx": {"pkg.installed": []}})},
        "files/nginx.conf": {"content": "server {}"},
    }


@pytest.fixture
def other_files():
    return {
        "top.sls": {"content": yaml.dump({"base": {"*": ["bar"]}})},
        "bar.sls": {"content": yaml.dump({"nginx": {"pkg.removed": []}})},
    }


@pytest.mark.parametrize("files_present", ("other_files",), indirect=True)
def test_something_else(s3, files_present):
    """ ... """
```

Reply:
Yeah this isn't really helping readability. If it were the same files repeated half a dozen times it would be a different story, but the tradeoff isn't worth it yet, in my opinion.